libcxx/test/std/algorithms/alg.nonmodifying/alg.contains/ranges.contains.pass.cpp

```diff
@@ -0,0 +1,237 @@
+      assert(ret);
+    }
+  }
+
+  { // no match
+    ValueT a[] = {13, 1, 21, 4, 5};
+    auto whole = std::ranges::subrange(Iter(a), Sent(Iter(a + 5)));
+    {
+      bool ret = std::ranges::contains(whole.begin(), whole.end(), 10);
+      assert(!ret);
+    }
+    {
+      bool ret = std::ranges::contains(whole, 10);
+      assert(!ret);
+    }
+  }
+
+  { // check that the projections are used
+    int a[] = {1, 9, 0, 13, 25};
+    {
+      bool ret = std::ranges::contains(a, a + 5, -13, [&](int i) { return i * -1; });
+      assert(ret);
+    }
+    {
+      auto range = std::ranges::subrange(a, a + 5);
+      bool ret   = std::ranges::contains(range, -13, [&](int i) { return i * -1; });
+      assert(ret);
+    }
+  }
+
+  { // check that std::string type works
+    const std::string str{"hello world"};
+    const std::string str1{"hi world"};
+    std::string a[] = {str1, str1, str, str1, str1};
+    auto whole =
+        std::ranges::subrange(forward_iterator(std::move_iterator(a)), forward_iterator(std::move_iterator(a + 5)));
+    {
+      bool ret = std::ranges::contains(whole.begin(), whole.end(), "hello world");
+      assert(ret);
+    }
+    {
+      bool ret = std::ranges::contains(whole, "hello world");
+      assert(ret);
+    }
+  }
+
+  { // check that non-continuous iterators work
+    std::vector<bool> whole{false, false, true, false};
+    {
+      bool ret = std::ranges::contains(whole.begin(), whole.end(), true);
+      assert(ret);
+    }
+    {
+      bool ret = std::ranges::contains(whole, true);
+      assert(ret);
+    }
+  }
+
+  { // check that non-continuous iterators(views::transform) work
+    int a[]            = {1, 2, 3, 4, 5};
+    auto square_number = a | std::views::transform([](int x) { return x * x; });
+    {
+      bool ret = std::ranges::contains(square_number.begin(), square_number.end(), 16);
+      assert(ret);
+    }
+    {
+      bool ret = std::ranges::contains(square_number, 16);
+      assert(ret);
+    }
+  }
+}
+
+constexpr bool test() {
+  types::for_each(types::type_list<char, long long>{}, []<class T> {
+    types::for_each(types::cpp20_input_iterator_list<T*>{}, []<class Iter> {
+      if constexpr (std::forward_iterator<Iter>)
+        test_iterators<Iter>();
+      test_iterators<Iter, sentinel_wrapper<Iter>>();
+      test_iterators<Iter, sized_sentinel<Iter>>();
+    });
+  });
+
+  { // count invocations of the projection
+    int a[]              = {1, 9, 0, 13, 25};
+    int projection_count = 0;
+    {
+      bool ret = std::ranges::contains(a, a + 5, 0, [&](int i) {
+        ++projection_count;
+        return i;
+      });
+      assert(ret);
+      assert(projection_count == 3);
+      projection_count = 0;
+    }
+    {
+      bool ret = std::ranges::contains(a, 0, [&](int i) {
+        ++projection_count;
+        return i;
+      });
+      assert(ret);
+      assert(projection_count == 3);
+    }
+  }
+
+  return true;
+}
+
+int main(int, char**) {
+  test();
```

Comment:
To make non-constexpr tests you'll need to add a new function called `test_nonconstexpr()` or similar.

Then you can call that function from main (but not from the static_assert, which is what requires it to be constexpr).